Improve error message on MySQL Flavor autodetect failing

#### Feature Description

MYSQL_FLAVOR autodetection can fail because no mysqld is present locally, or VT_MYSQL_ROOT has been set to a strange location.

We should change the panic error message to include both of these environment variables so that bug reports can be better understood.

This will be more useful than the current output:

```
panic: could not detect version from mysqld --version or MYSQL_FLAVOR [recovered]panic: 
	could not detect version from mysqld --version or MYSQL_FLAVORpanic: could not detect version from mysqld --version or MYSQL_FLAVOR [recovered]


	goroutine panic: could not detect version from mysqld --version or MYSQL_FLAVOR1
 [
runninggoroutine ]:
vitess.io/vitess/go/exit.doRecover1 [running(]:
0xa9c360vitess.io/vitess/go/exit.doRecover, 0xc8c160, (0xc8c1000xa9c360)
, 	0xc8c160/home/planetscale/go/src/vitess.io/vitess/go/exit/exit.go:, 920xc8c100 +)
0x1b8	
/home/planetscale/go/src/vitess.io/vitess/go/exit/exit.govitess.io/vitess/go/exit.Recover:(92)
 +0x1b8	
/home/planetscale/go/src/vitess.io/vitess/go/exit/exit.govitess.io/vitess/go/exit.Recover:(69)
 +0x48	
/home/planetscale/go/src/vitess.io/vitess/go/exit/exit.gopanic:69 +(0x480xa9c360
, panic0xc8c160)
(	0xa9c360/usr/local/go/src/runtime/panic.go, :0xc8c160513)
 +0x1b9
vitess.io/vitess/go/vt/mysqlctl.NewMysqld(0x12405e0, 0x2f)
	/home/planetscale/go/src/vitess.io/vitess/go/vt/mysqlctl/mysqld.go:122 +0x48e
vitess.io/vitess/go/vt/mysqlctl.CreateMysqldAndMycnf(0xc000000066, 0x0, 0x0, 0x42ce, 0x0, 0x0, 0x2f, 0xc000186430)
	/home/planetscale/go/src/vitess.io/vitess/go/vt/mysqlctl/cmd.go:	54/usr/local/go/src/runtime/panic.go +0x1d7:
513main.initCmd +0x1b9(
0xc000077680vitess.io/vitess/go/vt/mysqlctl.NewMysqld, 0xc000030120(, 0x12405e00x2, , 0x2f0x2)
, 0x0	, /home/planetscale/go/src/vitess.io/vitess/go/vt/mysqlctl/mysqld.go0x0:)
122 +	0x48e/home/planetscale/go/src/vitess.io/vitess/go/cmd/mysqlctl/mysqlctl.go
:vitess.io/vitess/go/vt/mysqlctl.CreateMysqldAndMycnf68 +(0x14b0xc000000065
, main.main0x0, (0x0)
	, /home/planetscale/go/src/vitess.io/vitess/go/cmd/mysqlctl/mysqlctl.go0x42cd:, 2530x0 +, 0x39e0x0
, 0x2f, 0xc000186420)
	/home/planetscale/go/src/vitess.io/vitess/go/vt/mysqlctl/cmd.go:54 +0x1d7
main.initCmd(0xc000077680, 0xc000030120, 0x2, 0x2, 0x0, 0x0)
	/home/planetscale/go/src/vitess.io/vitess/go/cmd/mysqlctl/mysqlctl.go:68 +0x14b
main.main()
	/home/planetscale/go/src/vitess.io/vitess/go/cmd/mysqlctl/mysqlctl.go:253 +0x39e
panic: could not detect version from mysqld --version or MYSQL_FLAVOR [recovered]
	panic: could not detect version from mysqld --version or MYSQL_FLAVOR

goroutine 1 [running]:
vitess.io/vitess/go/exit.doRecover(0xa9c360, 0xc8c160, 0xc8c100)
	/home/planetscale/go/src/vitess.io/vitess/go/exit/exit.go:92 +0x1b8
vitess.io/vitess/go/exit.Recover()
	/home/planetscale/go/src/vitess.io/vitess/go/exit/exit.go:69 +0x48
panic(0xa9c360, 0xc8c160)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
vitess.io/vitess/go/vt/mysqlctl.NewMysqld(0x12405e0, 0x2f)
	/home/planetscale/go/src/vitess.io/vitess/go/vt/mysqlctl/mysqld.go:122 +0x48e
vitess.io/vitess/go/vt/mysqlctl.CreateMysqldAndMycnf(0xc000000064, 0x0, 0x0, 0x42cc, 0x0, 0x0, 0x2f, 0xc00018e430)
	/home/planetscale/go/src/vitess.io/vitess/go/vt/mysqlctl/cmd.go:54 +0x1d7
main.initCmd(0xc000077680, 0xc000030120, 0x2, 0x2, 0x0, 0x0)
	/home/planetscale/go/src/vitess.io/vitess/go/cmd/mysqlctl/mysqlctl.go:68 +0x14b
main.main()
	/home/planetscale/go/src/vitess.io/vitess/go/cmd/mysqlctl/mysqlctl.go:253 +0x39e
```

--- a/go/vt/mysqlctl/capabilityset.go
+++ b/go/vt/mysqlctl/capabilityset.go
@@ -46,34 +46,10 @@ func (c *capabilitySet) hasMySQLUpgradeInServer() bool {
 func (c *capabilitySet) hasInitializeInServer() bool {
 	return c.isMySQLLike() && c.version.atLeast(serverVersion{Major: 5, Minor: 7, Patch: 0})
 }
-func (c *capabilitySet) hasMySQLxEnabledByDefault() bool {
-	return c.isMySQLLike() && c.version.atLeast(serverVersion{Major: 8, Minor: 0, Patch: 11})
-}
-func (c *capabilitySet) hasPersistConfig() bool {
-	return c.isMySQLLike() && c.version.atLeast(serverVersion{Major: 8, Minor: 0, Patch: 0})
-}
-func (c *capabilitySet) hasShutdownCommand() bool {
-	return (c.isMySQLLike() && c.version.atLeast(serverVersion{Major: 5, Minor: 7, Patch: 9})) || (c.isMariaDB() && c.version.atLeast(serverVersion{Major: 10, Minor: 0, Patch: 4}))
-}
-func (c *capabilitySet) hasBackupLocks() bool {
-	return c.isMySQLLike() && c.version.atLeast(serverVersion{Major: 8, Minor: 0, Patch: 0})
-}
-func (c *capabilitySet) hasDefaultUft8mb4() bool {
-	return c.isMySQLLike() && c.version.atLeast(serverVersion{Major: 8, Minor: 0, Patch: 0})
-}
-func (c *capabilitySet) hasSemiSyncEnabledByDefault() bool {
-	return c.isMariaDB() && c.version.atLeast(serverVersion{Major: 10, Minor: 3, Patch: 3})
-}
 
 // IsMySQLLike tests if the server is either MySQL
 // or Percona Server. At least currently, Vitess doesn't
 // make use of any specific Percona Server features.
 func (c *capabilitySet) isMySQLLike() bool {
 	return c.flavor == flavorMySQL || c.flavor == flavorPercona
-}
-
-// IsMariaDB tests if the server is MariaDB.
-// IsMySQLLike() and IsMariaDB() are mutually exclusive
-func (c *capabilitySet) isMariaDB() bool {
-	return c.flavor == flavorMariaDB
 }

--- a/go/vt/mysqlctl/mysqld.go
+++ b/go/vt/mysqlctl/mysqld.go
@@ -115,13 +115,40 @@ func NewMysqld(dbcfgs *dbconfigs.DBConfigs) *Mysqld {
 	version, getErr := getVersionString()
 	f, v, err := parseVersionString(version)
 
-	// Fallback if required
+	/*
+	 By default Vitess searches in vtenv.VtMysqlRoot() for a mysqld binary.
+	 This is usually the VT_MYSQL_ROOT env, but if it is unset or empty, it
+	 will substitute VtRoot(). See go/vt/env/env.go.
+
+	 A number of subdirs inside vtenv.VtMysqlRoot() will be searched, see
+	 func binaryPath() for context. If no mysqld binary is found (possibly
+	 because it is in a container or both VT_MYSQL_ROOT and VTROOT are set
+	 incorrectly), there will be a fallback to using the MYSQL_FLAVOR env
+	 variable.
+
+	 If MYSQL_FLAVOR is not defined, there will be a panic.
+
+	 Note: relying on MySQL_FLAVOR is not recommended, since for historical
+	 purposes "MySQL56" actually means MySQL 5.7, which is a very strange
+	 behavior.
+	*/
+
 	if getErr != nil || err != nil {
 		f, v, err = getVersionFromEnv()
 		if err != nil {
-			panic("could not detect version from mysqld --version or MYSQL_FLAVOR")
+			vtenvMysqlRoot, _ := vtenv.VtMysqlRoot()
+			message := fmt.Sprintf(`could not auto-detect MySQL version. You may need to set VT_MYSQL_ROOT so a mysqld binary can be found, or set the environment variable MYSQL_FLAVOR if mysqld is not available locally:
+	VT_MYSQL_ROOT: %s
+	VTROOT: %s
+	vtenv.VtMysqlRoot(): %s
+	MYSQL_FLAVOR: %s
+	`,
+				os.Getenv("VT_MYSQL_ROOT"),
+				os.Getenv("VTROOT"),
+				vtenvMysqlRoot,
+				os.Getenv("MYSQL_FLAVOR"))
+			panic(message)
 		}
-
 	}
 
 	log.Infof("Using flavor: %v, version: %v", f, v)

